### PR TITLE
Document `rustc_index::vec::IndexVec`

### PR DESCRIPTION
**Pull request by oriongonza**
Document a few of the methods.

Part of https://github.com/rust-lang/rust/issues/93792.

### CODE REVIEW

File: compiler/rustc_index/src/vec.rs [hunk 1]
```diff
@@ -17,20 +17,7 @@ use crate::{Idx, IndexSlice};
 /// While it's possible to use `u32` or `usize` directly for `I`,
 /// you almost certainly want to use a [`newtype_index!`]-generated type instead.
 ///
-/// This allows to index the IndexVec with the new index type:
-///
-/// ```
-/// use crate as rustc_index;
```
Comment (dev-ardi) [inline compiler/rustc_index/src/vec.rs]: I wasn't able to get this to work because I couldn't import this.

File: compiler/rustc_index/src/vec.rs [hunk 2]
```diff
@@ -12,24 +12,30 @@ use std::vec;
...
+/// This allows to index the IndexVec with the new index type
+///
+///
 /// [`newtype_index!`]: ../macro.newtype_index.html
 #[derive(Clone, PartialEq, Eq, Hash)]
 #[repr(transparent)]
 pub struct IndexVec<I: Idx, T> {
     pub raw: Vec<T>,
-    _marker: PhantomData<fn(&I)>,
+    _marker: PhantomData<I>,
```
Comment (compiler-errors) [inline compiler/rustc_index/src/vec.rs:28]: Why did you change this? This is not a documentation change.

File: compiler/rustc_index/src/vec.rs [hunk 3]
```diff
@@ -12,24 +12,30 @@ use std::vec;
...
 
 /// An owned contiguous collection of `T`s, indexed by `I` rather than by `usize`.
+/// Its purpose is to avoid mixing indexes.
 ///
 /// While it's possible to use `u32` or `usize` directly for `I`,
 /// you almost certainly want to use a [`newtype_index!`]-generated type instead.
 ///
+/// This allows to index the IndexVec with the new index type
+///
+///
```
Comment (compiler-errors) [inline compiler/rustc_index/src/vec.rs:22]:
```suggestion
```

File: compiler/rustc_index/src/vec.rs [hunk 4]
```diff
@@ -12,24 +12,30 @@ use std::vec;
...
 }
 
 impl<I: Idx, T> IndexVec<I, T> {
+    /// Constructs a new, empty `IndexVec<I, T>`.
     #[inline]
     pub const fn new() -> Self {
         IndexVec::from_raw(Vec::new())
     }
 
+    /// Constructs a new `IndexVec<I, T>` from a `Vec<T>`
```
Comment (compiler-errors) [inline compiler/rustc_index/src/vec.rs:38]: The usage of `.` is inconsistent between these three doc comments.

File: compiler/rustc_index/src/vec.rs [hunk 5]
```diff
@@ -59,6 +65,7 @@ impl<I: Idx, T> IndexVec<I, T> {
         IndexVec::from_raw(vec![elem; universe.len()])
     }
 
+    /// Creates a new IndexVec
```
Comment (compiler-errors) [inline compiler/rustc_index/src/vec.rs:68]: This doc comment seems incomplete

**Review by compiler-errors — CHANGES_REQUESTED** [review]

File: compiler/rustc_index/src/vec.rs (hunk 2 above)
Comment (dev-ardi) [reply compiler/rustc_index/src/vec.rs:28]: Oops sorry I miscomitted that.

File: compiler/rustc_index/src/vec.rs [hunk 6]
```diff
@@ -12,10 +12,12 @@ use std::vec;
 use crate::{Idx, IndexSlice};
 
 /// An owned contiguous collection of `T`s, indexed by `I` rather than by `usize`.
+/// Its purpose is to avoid mixing indexes.
 ///
 /// While it's possible to use `u32` or `usize` directly for `I`,
 /// you almost certainly want to use a [`newtype_index!`]-generated type instead.
 ///
+/// This allows to index the IndexVec with the new index type.
```
Comment (wesleywiser) [inline compiler/rustc_index/src/vec.rs:20]: I'm not entirely sure but I think you might need an empty line after this one to make the `[newtype_index!]` link resolve. 

```suggestion
/// This allows to index the IndexVec with the new index type.
///
```

**Review by wesleywiser — APPROVED** [review]

**Comment by wesleywiser** [comment]
Thanks @dev-ardi! Always nice to see docs improvements 👍 

@bors r+ rollup
